### ISSUE TITLE
SelectorBuilder doesn't check for errors properly

### ISSUE THREAD

**Issue by pavlosgi** [ghostdogpr/caliban#1162]
When using the generated SelectorBuilder the way it's checking for errors seems wrong or the server response is wrong. The two don't line up.

Server returns a response of the following shape when are error occurs:
 {"data":{"createUser":null},"errors":[{"message" ... 

So data itself isn't Null as the decode function checks in the SelectorBuilder. Is this a bug on the SelectorBuilder or should the server be returning  {"data":null,"errors":[{"message" instead. ...

Either way it seems one of the two needs to change, unless I am missing something.

```
  def decode(payload: String): Either[CalibanClientError, (A, List[GraphQLResponseError], Option[Json])] =
    for {
      parsed      <- parser
                       .decode[GraphQLResponse](payload)
                       .left
                       .map(ex => DecodingError("Json deserialization error", Some(ex)))
      data        <- if (parsed.errors.nonEmpty && parsed.data.forall(_ == __Value.__NullValue))
                       Left(ServerError(parsed.errors))
                     else Right(parsed.data)
      objectValue <- data match {
                       case Some(o: __ObjectValue) => Right(o)
                       case _                      => Left(DecodingError("Result is not an object"))
                     }
      result      <- self.fromGraphQL(objectValue)
    } yield (result, parsed.errors, parsed.extensions)
```

**Comment by paulpdaniels** [comment]
@pavlosgi it depends on your server resolver, GraphQL can return partial errors which is what you see here, the field that errored becomes null and the error is supplied in the errors array. You can make it fail the whole query as well if you want (depends on the implementation but if you are using caliban as the backend you have to make sure you are pushing a `Defect` out of your resolver for it to fail the whole query.

**Comment by ghostdogpr** [comment]
If your field on the server side is nullable and fails, the field only will return null and the rest of the query will be executed.
If your field on the server side is non-nullable and fails, the whole query will fail.

The first case is mapped to an `Option[A]` in Caliban Client, while the second case is mapped to `A`. If you want the whole query to fail, make the server field be nullable (e.g. changing a `Task` into a `UIO`, see https://ghostdogpr.github.io/caliban/docs/schema.html#schemas).

**Comment by ghostdogpr** [comment]
I should add that you can use `toRequestWith` instead of `toRequest` to access the partial errors in case of success.

**Comment by pavlosgi** [comment]
Thank you for your quick responses. I am not sure I understand why this is desirable here. It sounds like I need to turn my schema response type to Option so that the client happily parses that Option type when errors exist. I don't think I should have to do that.
So for example if I have:
a) a simple mutation createUser(email: String!): UserId! 
b) ZIO task that fails due to a bad email (so bad request)

Making UserId optional so that the above doesn't end up being a decoding failure seems to me like I am describing failure twice once in the Optional response type and once in the errors that will appear in the response. Given there are errors in the response shouldn't the client know to avoid trying to parse the UserId? Therefore making the actual response type Either[UserErrors, UserId] even though we aren't describing this fully in the graphql schema definition.

**Comment by ghostdogpr** [comment]
@pavlosgi your two bullet points a) and b) can't be true at the same time.

If your resolver field is like this: `createUser(email: EmailArgs): Task[UserId]`, the generated schema will be `createUser(email: String!): UserId` (not `UserId!`). That is because `Task[A]` is turned into a nullable A (see [this](https://ghostdogpr.github.io/caliban/docs/schema.html#schemas)). You can try printing `api.render` and you will notice that `UserId` is nullable.

If you don't want that, you need to change the field to `createUser(email: EmailArgs): UIO[UserId]` and call `.orDie` on your method when you create the resolver to turn `Task` into `UIO`. Then it will generate `createUser(email: String!): UserId!`. And an error will indeed fail the whole query, and the client request will be a failure.

In short, we use the 2 different error channels of ZIO to express the nullability: `UIO/URIO` for non-nullable and `Task`/`RIO` for nullable.

**Comment by pavlosgi** [comment]
I see that makes sense. Currently I am generating everything off of a gql schema. In my schema I have my mutation type be UserId! and I generate using the abstractEffectType. I just used orDie in my resolver and it all works now as you described. Thanks again for answering all my points.
